nemo_aligner/models/nlp/gpt/megatron_gpt_spin_model.py

```diff
@@ -400,18 +397,95 @@ def onload_adam_states(self):
             self.distributed_adam_offload_manager.__exit__(None, None, None)
 
         self.distributed_adam_offload_manager = None
-
-    def sharded_state_dict(self, prefix: str = ""):
-        sharded_state_dict = super().sharded_state_dict(prefix=prefix)
-
-        # add in the reference policy weights
+    
+    # Alternative 1: swap in the ref weights and create a custom key which then needs to be handled in the loader
+    def sharded_state_dict_alt_1(self, prefix: str = ''):
+        sharded_state_dict_orig = super().sharded_state_dict(prefix=prefix)
+        
+        if self.ref_policy_state_dict is None:
+            return sharded_state_dict_orig
+        
+        sharded_state_dict_new = {}
+        with cpu_weight_swap(self, self.ref_policy_state_dict, megatron_amp_O2=self.megatron_amp_O2):
+            module_prefix = f'{prefix}ref_policy_model.'
+            for index, module in enumerate(self.get_gpt_module_list()):
+                if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
+                    # virtual pipline rank must be set so that GPTModel returns the correct sharded state dict
+                    parallel_state.set_virtual_pipeline_model_parallel_rank(index)
+                    module_sharded_state_dict = module.sharded_state_dict(prefix=module_prefix)
+                    sharded_state_dict_new[f'ref_policy_model_{index}'] = module_sharded_state_dict
+                else:
+                    module_sharded_state_dict = module.sharded_state_dict(prefix=module_prefix)
+                    sharded_state_dict_new.update(module_sharded_state_dict)
+
+            # reset vp rank
+            if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
+                parallel_state.set_virtual_pipeline_model_parallel_rank(0)
+        
+        return sharded_state_dict_orig | sharded_state_dict_new
+    
+    # Alternative 2: use make_sharded_optimizer_tensor to link the TP/PP/DP info, but this may break under VP logic
```

Comment:
Maybe raise a `NotImplementedError` if VP is used, to be sure nobody accidentally uses it? (unless you know it crashes already -- I just want to make sure it won't lead to silent incorrect results)